cpp/src/arrow/status.h

```diff
@@ -83,6 +83,7 @@ enum class StatusCode : char {
   IOError = 5,
   CapacityError = 6,
   IndexError = 7,
+  Cancelled = 8,
```

Comment:
I'm not sure this is useful. Cancellations have an underlying cause, so perhaps force the source of cancellation to pass the actual cause?

Reply:
I think it's fine, think about one thread error-ing and the other receiving the cancel broadcast. It will be useful to distinguish which are real errors versus cancel-induced.